app/uk/gov/hmrc/serviceconfigs/persistence/DeploymentConfigSnapshotRepository.scala

```diff
@@ -66,16 +80,136 @@ class DeploymentConfigSnapshotRepository @Inject()(
       .toFuture
       .map(_ => ())
 
+  def removeLatestFlagForNonDeletedSnapshotsInEnvironment(environment: Environment): Future[Unit] =
+    collection
+      .updateMany(
+        filter = and(
+          equal("deploymentConfig.environment", environment.asString),
+          equal("latest", true),
+          equal("deleted", false),
+        ),
+        update = set("latest", false)
+      ).toFuture().map(_ => ())
+
+  def removeLatestFlagForServiceInEnvironment(serviceName: String, environment: Environment): Future[Unit] =
+    collection
+      .updateMany(
+        filter = and(
+          equal("deploymentConfig.name", serviceName),
+          equal("deploymentConfig.environment", environment.asString),
+        ),
+        update = set("latest", false)
+      ).toFuture().map(_ => ())
+
+
   def populate(date: Instant): Future[Unit] = {
 
     def forEnvironment(environment: Environment): Future[Unit] = {
       for {
         deploymentConfigs <- deploymentConfigRepository.findAll(environment)
-        snapshots         =  deploymentConfigs.map(DeploymentConfigSnapshot(date, _))
-        _                 <- collection.insertMany(snapshots).toFuture()
+        latestSnapshots   <- latestSnapshotsInEnvironment(environment)
+        planOfWork        =  PlanOfWork.fromLatestSnapshotsAndCurrentDeploymentConfigs(latestSnapshots.toList, deploymentConfigs.toList, date)
+        _                 <- executePlanOfWork(planOfWork, environment)
       } yield ()
     }
 
     Environment.values.foldLeftM(())((_, env) => forEnvironment(env))
   }
+
+  def executePlanOfWork(planOfWork: PlanOfWork, environment: Environment): Future[Unit] = {
+    logger.debug(s"Processing `DeploymentConfigSnapshot`s for ${environment.asString}")
+    val batchInsertions =
+      planOfWork.snapshots.map(_.deploymentConfigSnapshot) ++
+      planOfWork.snapshotSynthesisedDeletions.map(_.deploymentConfigSnapshot)
+
+    def reintroduceServiceSnapshot(snapshotServiceReintroduction: SnapshotServiceReintroduction) = {
+      logger.debug(s"Creating a snapshot for reintroduced service: ${snapshotServiceReintroduction.deploymentConfigSnapshot}")
+      for {
+        _ <- removeLatestFlagForServiceInEnvironment(
+          snapshotServiceReintroduction.deploymentConfigSnapshot.deploymentConfig.name,
+          snapshotServiceReintroduction.deploymentConfigSnapshot.deploymentConfig.environment
+        )
+        _ <- collection.insertOne(snapshotServiceReintroduction.deploymentConfigSnapshot).toFuture()
+      } yield ()
+    }
+
+    for {
+      _ <- if (batchInsertions.nonEmpty) removeLatestFlagForNonDeletedSnapshotsInEnvironment(environment) else Future.unit
+      _ <- collection.insertMany(batchInsertions).toFuture()
```

Comment:
if you inline `batchInsertions` here it keeps all the `planOfWork...` at the same level, making it easier to follow the logic